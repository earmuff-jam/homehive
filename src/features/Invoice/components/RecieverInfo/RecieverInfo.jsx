import { useEffect, useState } from "react";
import React from "react";

import { useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";

import dayjs from "dayjs";

import { Stack } from "@mui/material";
import CustomSnackbar from "common/CustomSnackbar/CustomSnackbar";
import RowHeader from "common/RowHeader/RowHeader";
import {
  useGetReceiverInfoQuery,
  useUpsertReceiverInfoMutation,
} from "features/Api/invoiceApi";
import UserInfoViewer from "features/Invoice/components/UserInfo/UserInfoViewer";
import { useAppTitle } from "hooks/useAppTitle";

export default function RecieverInfo() {
  useAppTitle("Reciever Information");
  const navigate = useNavigate();
  const [showSnackbar, setShowSnackbar] = useState(false);

  const {
    data: recieverInfo,
    isLoading: isRecieverInfoLoading,
    isSuccess: isRecieverInfoSuccess,
  } = useGetReceiverInfoQuery();

  const [
    upsertRecieverInfo,
    {
      isLoading: isUpsertRecieverInfoLoading,
      isSuccess: isUpsertRecieverInfoSuccess,
    },
  ] = useUpsertReceiverInfoMutation();

  const {
    register,
    handleSubmit,
    formState: { errors, isValid },
    reset,
  } = useForm({
    mode: "onChange",
    defaultValues: {
      first_name: "",
      last_name: "",
      email_address: "",
      phone_number: "",
      street_address: "",
      city: "",
      state: "",
      zipcode: "",
    },
  });

  const submit = (formData) => {
    formData["updatedOn"] = dayjs();
    upsertRecieverInfo(formData);
  };

  useEffect(() => {
    if (isUpsertRecieverInfoSuccess) {
      setShowSnackbar(true);
    }
  }, [isUpsertRecieverInfoLoading]);

  useEffect(() => {
    if (isRecieverInfoSuccess) {
      reset({
        first_name: recieverInfo.first_name,
        last_name: recieverInfo.last_name,
        email_address: recieverInfo.email_address,
        phone_number: recieverInfo.phone_number,
        street_address: recieverInfo.street_address,
        city: recieverInfo.city,
        state: recieverInfo.state,
        zipcode: recieverInfo.zipcode,
        updatedOn: recieverInfo.updatedOn,
      });
    }
  }, [isRecieverInfoLoading]);

  return (
    <Stack spacing={1} alignItems="center" data-tour={"reciever-0"}>
      <RowHeader
        title="Add details about the reciever"
        caption="Required fields are marked with an * "
      />
      <UserInfoViewer
        title="Reciever Information"
        caption="Add details about the reciever"
        register={register}
        errors={errors}
        isDisabled={!isValid}
        onSubmit={handleSubmit(submit)}
        loading={isUpsertRecieverInfoLoading}
      />
      <CustomSnackbar
        showSnackbar={showSnackbar}
        setShowSnackbar={setShowSnackbar}
        title="Changes saved."
        caption="View Invoice"
        onClick={() => navigate("/view")}
      />
    </Stack>
  );
}
