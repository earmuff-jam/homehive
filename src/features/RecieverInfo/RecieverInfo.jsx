import { useEffect, useState } from "react";
import UserInfoViewer from "../../common/UserInfo/UserInfoViewer";
import { BLANK_INDIVIDUAL_INFORMATION_DETAILS } from "../../common/UserInfo/constants";
import RowHeader from "../../common/RowHeader/RowHeader";
import { Stack } from "@mui/material";
import dayjs from "dayjs";
import CustomSnackbar from "../../common/CustomSnackbar/CustomSnackbar";
import { useNavigate } from "react-router-dom";

export default function RecieverInfo() {
  const navigate = useNavigate();
  const [showSnackbar, setShowSnackbar] = useState(false);
  const [formData, setFormData] = useState(
    BLANK_INDIVIDUAL_INFORMATION_DETAILS
  );

  const handleChange = (ev) => {
    const { id, value } = ev.target;
    const updatedFormData = { ...formData };
    let errorMsg = "";
    for (const validator of updatedFormData[id].validators) {
      if (validator.validate(value)) {
        errorMsg = validator.message;
        break;
      }
    }
    updatedFormData[id] = {
      ...updatedFormData[id],
      value,
      errorMsg,
    };
    setFormData(updatedFormData);
  };

  const submit = (ev) => {
    ev.preventDefault();
    const draftData = Object.entries(formData).reduce(
      (acc, [key, valueObj]) => {
        acc[key] = valueObj.value;
        return acc;
      },
      {}
    );
    draftData["updated_on"] = dayjs();
    localStorage.setItem("recieverInfo", JSON.stringify(draftData));
    setShowSnackbar(true);
  };

  const isDisabled = () => {
    const containsErr = Object.values(formData).reduce((acc, el) => {
      if (el.errorMsg) {
        return true;
      }
      return acc;
    }, false);

    const requiredFormFields = Object.values(formData).filter(
      (v) => v.isRequired
    );
    const isRequiredFieldsEmpty = requiredFormFields.some(
      (el) => el.value.trim() === ""
    );

    return containsErr || isRequiredFieldsEmpty;
  };

  useEffect(() => {
    const localValues = localStorage.getItem("recieverInfo");
    const parsedValues = JSON.parse(localValues);
    if (parsedValues) {
      const draftProfileDetails = { ...BLANK_INDIVIDUAL_INFORMATION_DETAILS };
      draftProfileDetails.first_name.value = parsedValues.first_name;
      draftProfileDetails.last_name.value = parsedValues.last_name;
      draftProfileDetails.email_address.value = parsedValues.email_address;
      draftProfileDetails.phone_number.value = parsedValues.phone_number;
      draftProfileDetails.street_address.value = parsedValues.street_address;
      draftProfileDetails.city.value = parsedValues.city;
      draftProfileDetails.state.value = parsedValues.state;
      draftProfileDetails.zipcode.value = parsedValues.zipcode;
      draftProfileDetails.updated_on = parsedValues.updated_on;
      setFormData(draftProfileDetails);
    }
  }, []);

  return (
    <Stack spacing={1} alignItems="center">
      <RowHeader
        title="Add details about the reciever"
        caption="Required fields are marked with an * "
      />
      <UserInfoViewer
        title="Reciever Information"
        caption="Add details about the reciever"
        formData={formData}
        handleChange={handleChange}
        onSubmit={submit}
        isDisabled={isDisabled()}
      />
      <CustomSnackbar
        showSnackbar={showSnackbar}
        setShowSnackbar={setShowSnackbar}
        title="Changes saved."
        caption="View Invoice"
        onClick={() => navigate("/view")}
      />
    </Stack>
  );
}
