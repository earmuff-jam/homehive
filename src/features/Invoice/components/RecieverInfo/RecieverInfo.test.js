import React from "react";

import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import RecieverInfo from "./RecieverInfo";
import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";

jest.mock("hooks/useAppTitle", () => ({
  useAppTitle: jest.fn(),
}));

jest.mock("common/RowHeader/RowHeader", () => ({
  __esModule: true,
  default: ({ title, caption }) => (
    <div>
      <h1>{title}</h1>
      <p>{caption}</p>
    </div>
  ),
}));

jest.mock("features/Invoice/components/UserInfo/UserInfoViewer", () => ({
  __esModule: true,
  default: ({ onSubmit }) => (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        onSubmit({ first_name: "John" });
      }}
    >
      <button type="submit">Submit</button>
    </form>
  ),
}));

jest.mock("features/Api/invoiceApi", () => ({
  useGetReceiverInfoQuery: () => ({
    data: {},
    isLoading: false,
    isSuccess: true,
  }),
  useUpsertReceiverInfoMutation: () => [
    jest.fn(),
    { isLoading: false, isSuccess: false },
  ],
}));

describe("RecieverInfo component", () => {
  it("renders correctly and matches snapshot", () => {
    const store = configureStore({ reducer: () => ({}) });

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <RecieverInfo />
        </MemoryRouter>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
