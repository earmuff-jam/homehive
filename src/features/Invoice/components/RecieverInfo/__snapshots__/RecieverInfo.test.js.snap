// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecieverInfo component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-aha7cr-MuiStack-root"
    data-tour="reciever-0"
  >
    <div>
      <h1>
        Add details about the reciever
      </h1>
      <p>
        Required fields are marked with an * 
      </p>
    </div>
    <form>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
